Enki Cleanup and Debugging

sparksam opened the pull request:
## Enki cleanup:
- [x] Moved `cnx-epub` code to `enki`
- [x] Remove `cnx-epub` dependency
- [x] Added annotation to time functions execution time for future algorithm optimization
- [x] Added python debugging inside the docker container
## TODO
- [x] Add Java debugging inside Enki container
- [x] Add Javascript debugging inside Enki container

philschatz commented on bakery-js/package.json:
```diff
@@ -15,7 +15,8 @@
     "test": "jest --coverage",
-    "test:ci": "jest --ci --maxWorkers=2 --reporters=default --reporters=jest-junit"
+    "test:ci": "jest --ci --maxWorkers=2 --reporters=default --reporters=jest-junit",
+    "watch": "npx nodemon --inspect=0.0.0.0:9229 --watch dist/index.js "
```
Is this called by anything? I didn't find a reference to it anywhere. 

I haven't used nodemon in a long time but does it provide anything more than `node --inspect=0.0.0.0:9229`?

If it is being used for development then it might be good to include nodemon in the list of devDependencies.

philschatz commented on bakery-src/scripts/cnx_formatters.py:
```diff
@@ -0,0 +1,74 @@
+    def __str__(self):  # pragma: no cover
+        if IS_PY3:
+            return self.__bytes__().decode('utf-8')
+        return self.__bytes__()
```
It might be a good time to remove the Python 2 code

philschatz commented on dockerfiles/docker-entrypoint.sh:
```diff
@@ -93,15 +95,14 @@ function do_xhtml_validate() {
+    echo "Validating xhtml links in $dir_name"
+    if [[ $JAVA_DEBUG == 1 ]]
+    then
+        extra_vars="-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=${DEBUG_HOST}:${JAVA_DEBUG_PORT}"
```
Ending the line with `# LCOV_EXCL_LINE` should fix the coverage

philschatz commented on dockerfiles/steps/step-epub.bash:
```diff
@@ -6,7 +6,13 @@ set -Eeuxo pipefail
 
-node --unhandled-rejections=strict "$JS_UTILS_STUFF_ROOT/bin/bakery-helper" epub ./ "$IO_EPUB/"
+extra_vars=""
+(( JS_DEBUG==1 )) && {
```
Similarly `# LCOV_EXCL_START` should fix this

philschatz replied:
Also, rather than remembering to add this block everywhere node is called (like in mathify, the docx conversion, adding source information while assembling, ...), it [seems](https://stackoverflow.com/a/71688732) an environment variable can be set in a common place like the docker-entrypoint:

```
NODE_OPTIONS='--inspect-brk=0.0.0.0:9229'
```

philschatz commented on dockerfiles/steps/step-epub.bash:
```diff
@@ -6,7 +6,13 @@ set -Eeuxo pipefail
+extra_vars=""
+(( JS_DEBUG==1 )) && {
+    echo "Javascript debug mode is on"
+    extra_vars+="--inspect-brk=0.0.0.0:9229"
```
Other pars use `DEBUG_HOST` and `##_DEBUG_PORT`. Any reason this is different?

Also, it seems like this is missing from other node steps like mathify

sparksam replied:
Will update this part.

philschatz reviewed (commented):
Ooh, I'm excited to try these debuggers out!

I haven't tried to debug yet but reading the code brought up a couple questions:

1. Was updating `styles` and `cookbook` submodules intentional?
2. The mathify code in `step-pdf` does not seem to be debuggable
3. The "Annotating XML files" code in `step-prebake` does not seem to be debuggable

Also, I'm curious what others think about:
1. Adding code to enable the python debugger
2. Having separate commandline flags to enable the different debuggers

philschatz commented on README.md:
```diff
@@ -203,6 +211,26 @@ START_AT_STEP=git-bake ./enki ./data/tin-bk all-pdf
+
+### Javascript Debugging
+
+In order to debug the Javascript code, you can set the breakpoints in the Javascript code under `bakery-js/src` and then run the `./enki` command with the `--jsdb` flag. The node program will then suspend till the client is connected to it and stop at the breakpoints and you can step through the code.
```
I was unable to set breakpoints in files within `bakery-js/src/`.

I was however able to set breakpoints in `bakery/dist/` after running `npm install && npm run build` inside bakery-js. Maybe there is a way (via commandline) to tell the node process to use sourcemaps when given breakpoints?

sparksam commented:
> Ooh, I'm excited to try these debuggers out!
> 
> I haven't tried to debug yet but reading the code brought up a couple questions:
> 
> 1. Was updating `styles` and `cookbook` submodules intentional?
> 2. The mathify code in `step-pdf` does not seem to be debuggable
> 3. The "Annotating XML files" code in `step-prebake` does not seem to be debuggable
> 4. Was there no gogod way to get the ruby debugger working?
> 
> Also, I'm curious what others think about:
> 
> 1. Adding code to enable the python debugger
> 2. Having separate commandline flags to enable the different debuggers

1. `cookbook` and `styles` update was not intentional. 
2. I added a new debugging entry to the `launch.json` for `mathify` because I couldn't figure out how to map multiple source paths
3. I updated the command for the `pre-bake` step
4. For Ruby 3 I was able to find [ways](https://github.com/ruby/debug) to debug remotely but couldn't figure for `byebug` inside the container. Will keep looking.

philschatz commented on .vscode/launch.json:
```diff
@@ -0,0 +1,53 @@
+      "skipFiles": ["<node_internals>/**"]
+    },
+    {
+      "name": "Mathify: Remote Debug",
```
I was able to debug all the JS/TS files by doing [this](https://github.com/openstax/enki/compare/sk-ce-1953...phil-ce-1953) (set a `NODE_OPTIONS` environment var and edit `.vscode/launch.json`)

sparksam replied:
Thanks, @philschatz. I'll pull it to my branch and test it.

philschatz reviewed (commented):
I don't think that having a separate Bakery-JS and Mathify debugger is necessary. Also, using the NODE_OPTIONS environment variable removes the need to explicitly find all the places where node is called

philschatz commented:
Oops, I missed being able to debug 1 javascript file: `mml2svg2png-json-rpc.js`.

Changing the call in `step-docx.bash` to something like this should fix it ([source](https://stackoverflow.com/q/29900253)):

```sh
"$BAKERY_SCRIPTS_ROOT/scripts/node_modules/.bin/pm2" start mml2svg2png-json-rpc.js --node-args="-r esm $NODE_OPTIONS" --wait-ready --listen-timeout 8000
```

philschatz reviewed (approved):
looks good! If you want, the `mml2svg2png-json-rpc.js` can be a separate PR.

I'm also excited because this PR should also unlock #197 (Mac M1) because cnx-epub is pinned to an an old version of lxml that does not work in the new version of Ubuntu
